be/src/exec/stream/aggregate/agg_state_data.cpp

```diff
@@ -0,0 +1,312 @@
+            }
+#endif
+            _agg_function->restore_detail(_agg_fn_ctx, i, columns, agg_group_state[i] + _agg_state_offset);
+        } else {
+            auto status = result_chunk_or.status();
+            DCHECK(status.is_end_of_file());
+            if (!status.is_end_of_file()) {
+                return status;
+            }
+        }
+    }
+    DCHECK(j == result_chunks.size());
+    return Status::OK();
+}
+
+Status AggStateData::process_chunk(size_t chunk_size, const std::vector<ChunkPtrOr>& result_chunks,
+                                   const Buffer<uint8_t>& keys_not_in_map, const StreamRowOp* ops,
+                                   std::vector<std::vector<const vectorized::Column*>>& raw_columns,
+                                   const Buffer<vectorized::AggDataPtr>& agg_group_state) const {
+    auto* columns = (raw_columns[agg_func_id()]).data();
+    for (int i = 0; i < chunk_size; i++) {
+        auto op = ops[i];
+        switch (op) {
+        case StreamRowOp::INSERT:
+        case StreamRowOp::UPDATE_BEFORE:
+            _agg_function->update(_agg_fn_ctx, columns, agg_group_state[i] + _agg_state_offset, i);
+            break;
+        case StreamRowOp::DELETE:
+        case StreamRowOp::UPDATE_AFTER:
+            _agg_function->retract(_agg_fn_ctx, columns, agg_group_state[i] + _agg_state_offset, i);
+            break;
+        }
+    }
+    return Status::OK();
+}
+
+Status AggStateData::finalize_result(size_t chunk_size, const Buffer<vectorized::AggDataPtr>& agg_group_data,
+                                     vectorized::Column* to) const {
+    _agg_function->batch_finalize(_agg_fn_ctx, chunk_size, agg_group_data, _agg_state_offset, to);
+    return Status::OK();
+}
+
+Status AggStateData::finalize_detail_result(size_t chunk_size, const vectorized::Columns& group_by_columns,
+                                            const Buffer<vectorized::AggDataPtr>& agg_group_data,
+                                            const StateTable* detail_state_table, vectorized::Column* to) const {
+    if (is_detail_state_table()) {
+        DCHECK(detail_state_table);
+        UInt8ColumnPtr is_sync_col = vectorized::UInt8Column::create();
+        for (size_t i = 0; i < chunk_size; i++) {
+            _agg_function->output_is_sync(_agg_fn_ctx, chunk_size, is_sync_col.get(),
+                                          agg_group_data[i] + _agg_state_offset);
+        }
+
+        uint8_t* is_sync_data = is_sync_col->mutable_raw_data();
+        // if need sync, query data from detail table.
+        for (size_t i = 0; i < chunk_size; i++) {
+            if (!is_sync_data[i]) {
+                continue;
+            }
+            DatumRow keys;
+            for (auto j = 0; j < group_by_columns.size(); j++) {
+                keys.push_back(group_by_columns[j]->get(i));
+            }
+            auto chunk_iter_or = detail_state_table->prefix_scan(keys);
```

Comment:
How to optimize it into a batch interface?

Reply:
Good question. `prefix_scan ` will output a `chunk_iter_or ` which may contain multi rows, here not use a batch interface first. Maybe optimize it later.